pkg/runtime/wfengine/activity.go

```diff
@@ -0,0 +1,298 @@
+		SequenceNumber: int64(taskEvent.EventId),
+		InstanceID:     api.InstanceID(workflowID),
+		NewEvent:       taskEvent,
+		Properties:     make(map[string]interface{}),
+	}
+
+	// Executing activity code is a one-way operation. We must wait for the app code to report its completion, which
+	// will trigger this callback channel.
+	// TODO: Need to come up with a design for timeouts. Some activities may need to run for hours but we also need
+	//       to handle the case where the app crashes and never responds to the workflow. It may be necessary to
+	//       introduce some kind of heartbeat protocol to help identify such cases.
+	callback := make(chan bool)
+	wi.Properties[CallbackChannelProperty] = callback
+	if err = a.scheduler.ScheduleActivity(ctx, wi); err != nil {
+		if errors.Is(err, context.DeadlineExceeded) {
+			return newRecoverableError(errors.New(
+				"timed-out trying to schedule an activity execution - this can happen if too many activities are running in parallel or if the workflow engine isn't running"))
+		}
+		return newRecoverableError(fmt.Errorf("failed to schedule an activity execution: %w", err))
+	}
+
+loop:
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-time.After(10 * time.Minute):
+			if deadline, ok := ctx.Deadline(); ok {
+				wfLogger.Warnf("%s: '%s' is still running - will keep waiting until %v", actorID, name, deadline)
+			} else {
+				wfLogger.Warnf("%s: '%s' is still running - will keep waiting indefinitely", actorID, name)
+			}
+		case completed := <-callback:
+			if completed {
+				break loop
+			} else {
+				return newRecoverableError(errExecutionAborted)
+			}
+		}
+	}
+
+	// publish the result back to the workflow actor as a new event to be processed
+	resultData, err := backend.MarshalHistoryEvent(wi.Result)
+	if err != nil {
+		return err
+	}
+	req := invokev1.
+		NewInvokeMethodRequest(AddWorkflowEventMethod).
+		WithActor(WorkflowActorType, workflowID).
+		WithRawDataBytes(resultData).
+		WithContentType(invokev1.OctetStreamContentType)
+	defer req.Close()
+
+	resp, err := a.actorRuntime.Call(ctx, req)
+	if err != nil {
+		return newRecoverableError(fmt.Errorf("failed to invoke '%s' method on workflow actor: %w", AddWorkflowEventMethod, err))
+	}
+	defer resp.Close()
+	return nil
+}
+
+// InvokeTimer implements actors.InternalActor
+func (*activityActor) InvokeTimer(ctx context.Context, actorID string, timerName string, params []byte) error {
+	return nil // timers are not used
+}
+
+// DeactivateActor implements actors.InternalActor
+func (a *activityActor) DeactivateActor(ctx context.Context, actorID string) error {
+	wfLogger.Debugf("deactivating activity actor '%s'", actorID)
+	a.statesCache.Delete(actorID)
+	return nil
+}
+
+func (a *activityActor) loadActivityState(ctx context.Context, actorID string, generation uint64) (activityState, error) {
+	// See if the state for this actor is already cached in memory.
+	result, ok := a.statesCache.Load(actorID)
```

Comment:
I'm assuming that due to the actor's nature of ensuring activation isolation across actors of the same ID you don't need to handle concurrent requests here - e.g using a RW mutex, is that correct?

Reply:
While it's true that we get the guarantees that any specific actor ID doesn't have concurrent execution, this `activityActor` struct is shared for all actor instances, so we do need to be careful about concurrency. The `a.statesCache` field is a `sync.Map`, though, so we should be safe for concurrent execution here.

Reply:
It means that we can have concurrently executions, even using a sync.Map you can have racing conditions where both tries to load from the map at exactly the same time, both returning false and proceeding with the method execution.

A https://pkg.go.dev/golang.org/x/sync/singleflight would be a good choice for this. Using a singleflight with the actor id as a key should solve the problem

Reply:
The guarantees of actors ensures that we'll never have two concurrent operations with the same `actorID` value, so we don't need to worry about to requests for the same actor racing with each other. The only thing we need to protect against is operations for two _different_ actors causing a panic due to concurrent map modifications. `sync.Map` should give us all the protection we need for that.

Reply:
Thanks for the clarification.